### ISSUE TITLE
Tutorial for running crypten in multiple hosts

### ISSUE THREAD

**Issue by thanksunix** [facebookresearch/CrypTen#104]
Hi guys,

Is there a tutorial for running Crypten examples in real multiple hosts instead of simulated process . I wanna compare the performance with TFE when having bandwidth and latency constraint.

**Comment by vshobha** [comment]
Hi @thanksunix 

Thanks for trying out CrypTen! Tutorial 6 shows you how to launch CrypTen on multiple AWS nodes using our helper scripts in CrypTen/scripts/  Let us know if you have any issues when trying them out!

**Comment by thanksunix** [comment]
> Hi @thanksunix
> 
> Thanks for trying out CrypTen! Tutorial 6 shows you how to launch CrypTen on multiple AWS nodes using our helper scripts in CrypTen/scripts/ Let us know if you have any issues when trying them out!

Thanks for reply.  @vshobha 
I dont have AWS nodes available. I wonder is there a tutorial for running them on existing machines? Thanks.

**Comment by knottb** [comment]
If you are using AWS, we have an Tutorial on launching jobs across multiple AWS instances [here](https://github.com/facebookresearch/CrypTen/blob/master/tutorials/Tutorial_6_CrypTen_on_AWS_instances.ipynb).


Assuming you are not using AWS instances, the process is pretty similar. We need to be able to do 2 things:

(1) You need some way of connecting your nodes
(2) You need to run your script on both nodes


For (1), we use environment variables ([default values](https://github.com/facebookresearch/CrypTen/blob/master/crypten/communicator/distributed_communicator.py#L76-L82)) to input into torch.distributed. See [this](https://pytorch.org/docs/stable/distributed.html#tcp-initialization) for information on how to connect. The environment variables you will likely need to provide are:

WORLD_SIZE   (number of nodes)
RANK                (rank of the current node (0 to WORLD_SIZE - 1))
RENDEZVOUS (init_method for torch.distributed)

Once these are set, you can just run your script on each of your nodes.

**Comment by thanksunix** [comment]
> If you are using AWS, we have an Tutorial on launching jobs across multiple AWS instances [here](https://github.com/facebookresearch/CrypTen/blob/master/tutorials/Tutorial_6_CrypTen_on_AWS_instances.ipynb).
> 
> Assuming you are not using AWS instances, the process is pretty similar. We need to be able to do 2 things:
> 
> (1) You need some way of connecting your nodes
> (2) You need to run your script on both nodes
> 
> For (1), we use environment variables ([default values](https://github.com/facebookresearch/CrypTen/blob/master/crypten/communicator/distributed_communicator.py#L76-L82)) to input into torch.distributed. See [this](https://pytorch.org/docs/stable/distributed.html#tcp-initialization) for information on how to connect. The environment variables you will likely need to provide are:
> 
> WORLD_SIZE (number of nodes)
> RANK (rank of the current node (0 to WORLD_SIZE - 1))
> RENDEZVOUS (init_method for torch.distributed)
> 
> Once these are set, you can just run your script on each of your nodes.

Thanks @knottb . This one works. :)

**Comment by krantikiran** [comment]
Hi everyone, 

I am trying to run a simple sigmoid function over two machines. (With IP addresses IP0 and IP1)
I set the environment variables to
WORLD_SIZE=2
RANDEZVOUS=tcp://IP0:PORT
RANK=0//(or 1)

However, the code hangs on crypten.init() when I use these variables.

The function is:
```
crypten.mpc.set_default_provider(crypten.mpc.provider.TrustedFirstParty)
crypten.init()
crypten.encoder.set_default_precision(12)
crypten.debug.configure_logging()

def run(rank, world_size):
    x_alice = crypten.load_from_party("alice_data", src=0)
    x_bob = crypten.load_from_party("bob_data", src=1)

    x_crypt = crypten.cat([x_alice, x_bob], dim=0)

    x_dec = crypten.common.approximations.sigmoid(x_crypt) 

    comm.get().print_communication_stats()
    print(len(x_dec.get_plain_text()))
    print(x_dec.get_plain_text())


run(0, 2)
```
python==3.7.9
torch==1.7+cpu
CrypTen at commit  2f6a7c5f20c10d3e55dc4a56a76da7640a0dad4c 

To tun the script, I am using
`python test.py --multiprocess`

**Comment by gmuraru** [comment]
I think that you need to run this code on both machines.
There is a ```distributed_launcher.py``` in scripts that might come in handy.

**Comment by krantikiran** [comment]
Thank you for the quick reply. 

I am running the code on both machines. 
Running the code using distributed_launcher.py is running both the processes on a single machine. 
It looked as if the RANDEZVOUS variable is being reset in distributed_launcher.py https://github.com/facebookresearch/CrypTen/blob/2f6a7c5f20c10d3e55dc4a56a76da7640a0dad4c/scripts/distributed_launcher.py#L34

**Comment by gmuraru** [comment]
I will take a look into this and will come back with what I found.

**Comment by krantikiran** [comment]
Thank you so much.

**Comment by gmuraru** [comment]
I looked into the problem! 

I think it might be only a typo: ```RANDEZVOUS``` --> ```RENDEZVOUS```

**Comment by gmuraru** [comment]
@krantikiran did it solved the problem? :D

**Comment by krantikiran** [comment]
Thank you for pointing that out. However, correcting the typo still didn't help.

**Comment by gmuraru** [comment]
If it is still blocking at ```crypten.init```, could you also try to set the ```GLOO_SOCKET_IFNAME``` to the interface name you are using for the connection?
Example:
```
export GLOO_SOCKET_IFNAME=="wlp59s0"
```

You can find the interface using ```ip a``` or ```ifconfig```

**Comment by krantikiran** [comment]
Even this option is not doing the job.

**Comment by gmuraru** [comment]
Still blocked at the ```crypten.init``` instruction, right? does any party show an error message?

**Comment by krantikiran** [comment]
There are no error messages. The party with RANK==1 times out after some time.

**Comment by gmuraru** [comment]
Could you join the CrypTen channel and ping me? we can do a more live debugging session there.
The instructions to join the slack channel should be at the end of the [README](https://github.com/facebookresearch/CrypTen/blob/master/README.md).

**Comment by hobbitlzy** [comment]
> I think that you need to run this code on both machines. There is a `distributed_launcher.py` in scripts that might come in handy.

Hi, I am confused about the function  of  `distributed_launcher.py`. I thought it simulates parties on one single machine using multiple processes, just like  `[MultiProcessLauncher](https://github.com/facebookresearch/CrypTen/blob/efe8edad571be1c586d0d9cefc562d562d4e9aa1/examples/multiprocess_launcher.py#L16)` do.

**Comment by zjTek** [comment]
Hi，could you tell me how to run crypten in mutiple computer that in the same local area network 

发自我的iPhone

**Comment by lvdmaaten** [comment]
@zjTek It would work the same as starting a distributed training job in PyTorch. You just launch a process running your CrypTen code on each machine, and you make sure the network is setup in such a way that enables rendezvous and communication between the machines.

**Comment by jimouris** [comment]
@krantikiran I'm having the same issue, did you figure it out?
